### PR TITLE
Disallow multivalued variables in visualizations

### PR DESCRIPTION
**Pull request by chowington**
Fixes https://github.com/VEuPathDB/web-eda/issues/279

Note that the default is to not allow multivalued variables

### CODE REVIEW

File: src/lib/core/utils/data-element-constraints.ts [hunk 1]
```diff
@@ -108,10 +111,14 @@ export function flattenConstraints(
...
       // If a value (variable) has not been user-selected for this constraint, then it is considered to be "in-play"
       if (value == null) return true;
-      // If a constraint does not declare shapes or types, then any value is allowed, thus the constraint is "in-play"
-      if (isEmpty(constraint.allowedShapes) || isEmpty(constraint.allowedTypes))
+      // If a constraint does not declare shapes or types and it allows multivalued variables, then any value is allowed, thus the constraint is "in-play"
+      if (
+        isEmpty(constraint.allowedShapes) &&
+        isEmpty(constraint.allowedTypes) &&
+        constraint.allowMultiValued
+      )
```
Comment (chowington) [inline src/lib/core/utils/data-element-constraints.ts:119]: Do review this please--I thought the existing ORing logic didn't match the comment/context

File: src/lib/core/utils/data-element-constraints.ts [hunk 2]
```diff
@@ -108,10 +111,14 @@ export function flattenConstraints(
...
         return true;
```
Comment (chowington) [inline src/lib/core/utils/data-element-constraints.ts:120]: Do review this please--I thought the existing ORing logic didn't match the comment/context

**Comment by bobular** [comment]
Any idea which datasets have multivalued variabels?
I just grepped through the entire GEMS1 study JSON and didn't find one!

**Comment by bobular** [comment]
Found some info by searching Slack
https://epvb.slack.com/archives/C012ZK4Q5CZ/p1629224907088700?thread_ts=1629144396.078600&cid=C012ZK4Q5CZ 

PRISM2 has some

- Plasmodium haplotype ID
- Non-malaria medication

**Review by bobular — APPROVED** [review]
I tested with PRISM2 and it worked.
Studied the code and it makes sense as far as I can see!

**Review by jtlong3rd — APPROVED** [review]

File: src/lib/core/utils/data-element-constraints.ts (hunk 2 above)
Comment (dmfalke) [reply src/lib/core/utils/data-element-constraints.ts:120]: Yep, good catch! It's a case of English "or" not matching semantics of logical "or" :slightly_smiling_face:

**Review by dmfalke — APPROVED** [review]

File: src/lib/core/utils/data-element-constraints.ts (hunk 2 above)
Comment (bobular) [reply src/lib/core/utils/data-element-constraints.ts:120]: Oh, I've only just seen that you changed the `||` to `&&` but that doesn't change the fact that I think what you have now is correct.

File: src/lib/core/utils/data-element-constraints.ts [hunk 3]
```diff
@@ -19,7 +19,9 @@ export function filterVariablesByConstraint(
 ): StudyEntity {
   if (
     constraint == null ||
-    (constraint.allowedShapes == null && constraint.allowedTypes == null)
+    (constraint.allowedShapes == null &&
+      constraint.allowedTypes == null &&
+      constraint.allowMultiValued)
```
[the remark this answers is not in the document]
Comment (chowington) [reply src/lib/core/utils/data-element-constraints.ts:24]: Whoops, forgot to add that comment. Sorry Bob! It would be helpful 😅

**Pull request merged**